### PR TITLE
[JENKINS-34670] Add support for a new full screen layout

### PR DESCRIPTION
**Pull request by recena**
[JENKINS-34670](https://issues.jenkins-ci.org/browse/JENKINS-34670)

- [x] Modify the Setup Wizard for using the full screen layout as a clear use-case.

The About page using the three different layout types (`1024px` x `768px`).
### full-screen: no header, no footer, no spaces

captura el 2016-07-10 a las 21 12 26
### one-column: full width

captura el 2016-07-10 a las 21 13 03
### two-column: side-panel + main-panel

captura el 2016-07-10 a las 21 14 08

@reviewbybees

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Looks good to me. 👍 and 🐝 
By the way, it may a bit affect users, who bundle custom Javascripts to headers and footers. Maybe makes sense to keep empty header and footer tags then

**Comment by recena** [comment]
@oleg-nenashev This kind of things should not do. This layout is thought specially for:

- Dashboard
- To replace the page when Jenkins is restarting, and avoid the side-effect of resource not loaded
- Maybe the InstallWizard (@kzantow)
- etc...

**Comment by oleg-nenashev** [comment]
@reviewbybees done BTW
@jenkinsci/code-reviewers I would appreciate additional feedback

**Comment by daniel-beck** [comment]
:-1: until the following questions are answered:

* Isn't that what `/lib/layout/html` does? Why not just use that?
* There's no rationale for this change. Why no headers/footers? I get that we may not want a side panel, but nothing?
* How is one supposed to navigate away from such pages? The example *About* page shows it's a dead end with no way to return, seems like a messy, unusable approach.

**Comment by oleg-nenashev** [comment]
Back to needs-review after the comment from @daniel-beck

**Comment by kzantow** [comment]
@daniel-beck `/lib/layout/html` should (and will) be removed in favor of this approach. the about page would use the one-column view, the no header/footer is for illustrative purposes only, which would replace the `/lib/layout/html`

:bee: from me on this change

**Comment by daniel-beck** [comment]
@kzantow 

> should … be removed in favor of this approach.

There's still no rationale for this change besides asserting it's the right thing to do. As I wrote,

> piling on options in lib/layout for exceptional pages like those without headers/footers [may] result in maintenance problems later

**Comment by recena** [comment]
@reviewbybees done

**Comment by recena** [comment]
@daniel-beck It seems there are 3 persons, included the person who added the `html.jelly`, agree with this is the right way to implement several layouts (included the full screen).

**Comment by recena** [comment]
If I provided a way to include support for several layout types, was precisely to avoid hacks that we can see in Jenkins core and plugins.

**Comment by daniel-beck** [comment]
@recena 

> there are 3 persons [who agree] this is the right way to implement several layouts

And yet none of them managed to explain _why_ it's a better approach, or even just addressed my specific concern above. There's only repeated assertions (another one from you just now) that this is the correct approach.

**Comment by kzantow** [comment]
@daniel-beck it's a better approach because it consolidates the style/script includes and resource references to a single file, instead of having to update 2 any time we might change them. And plugins could use this instead of defining local style overrides to hide the stuff, there's a common method of doing so.

**Comment by daniel-beck** [comment]
@kzantow 

Thank you! I feel we're getting closer to a result.

> it consolidates the style/script includes and resource references to a single file

This PR does not remove `html.jelly`, so there are no such benefits.

> plugins could use this instead of defining local style overrides to hide the stuff

They can use `html.jelly` -- they may even already do so, because nowhere does it state that it's not fit for use from plugins.

So, to recap:

* All the benefits of this change only apply if `html.jelly` gets removed.
* `html.jelly` doesn't get removed.
* In fact, it may be not possible to remove it, as it's effectively part of Jenkins' public plugin API now.

**Comment by kzantow** [comment]
@daniel-beck so, we could modify `html.jelly` to include the `layout.jelly` with the appropriate parameter to make it equivalent, probably remove the `html.jelly` usage in the setup wizard, and then figure out a way to deprecate it later.

**Comment by recena** [comment]
@daniel-beck 

This is a step to remove `html.helly` and provide a right way to provide full-screen layout. This step belongs to a more general solution that started few months ago.

**Comment by recena** [comment]
@kzantow Thanks for your assistance. My internet connection is very limited on these days.

**Comment by daniel-beck** [comment]
@kzantow Good idea. I think this should be fine if it included that change so there's actually a benefit to merging this.

**Comment by oleg-nenashev** [comment]
So it's ready to merge from what I see

**Comment by recena** [comment]
@oleg-nenashev From my side it is ready to merge but I understand @daniel-beck wants to see a real usage of that. My recommendation is to prepare a new PR.

**Comment by daniel-beck** [comment]
@oleg-nenashev @recena  I'd like to see actual applications of this code included in this same PR. Too much unused infrastructure in Jenkins as-is.

**Comment by recena** [comment]
@daniel-beck If I do it, this PR will be more difficult to review. The use cases are clear and were explained in this PR.

> Too much unused infrastructure in Jenkins as-is.

What do you mean? Examples?

In the UI, any cleanup or attempt to unify ways to do the things, should be considered. I don't know how this could be considered an `unused infrastructure`, especially when one of its goals it is precisely to remove code (I'd say _hacks_) that is not usable.

**Comment by daniel-beck** [comment]
@recena

> What do you mean? Examples?

How about #1788 which was implemented for basically the same purpose that is now served by #2323 and has no known applications? Once we add something, we cannot easily remove it anymore, so we should make sure we're adding the right thing, in the right way. 

There are no views actually making use of this change. Make it a new commit, or a separate PR, or a PR to a plugin if you're worried about reviewability, but do _something_ with the new layout you're adding here.

> any cleanup or attempt to unify ways to do the things, should be considered.

There's no cleanup here. That's the problem. You're just getting a mop and bucket but don't do anything with them, so they just clutter up the room.

**Comment by recena** [comment]
@daniel-beck 

> There's no cleanup here. That's the problem. You're just getting a mop and bucket but don't do anything with them, so they just clutter up the room.

Not everyone sees the same dirt 😄

**Comment by recena** [comment]
I think this PR is ready to be merge.

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers Seems some voting is required

**Comment by daniel-beck** [comment]
:-1: while there's no application for this. We carry around too much crap that's not used by anyone already

**Comment by recena** [comment]
@daniel-beck Ok.

more feedback is welcome @jenkinsci/code-reviewers

**Comment by recena** [comment]
@oleg-nenashev IMO, the label `needs-fix` is not correct. Nothing is needed to fix.

**Comment by recena** [comment]
@daniel-beck How it works in this OSS project? I need to get votes?

**Comment by recena** [comment]
@daniel-beck I'll send a new PR where you could see the usage. I'm trying to avoid a *big* PR.

**Comment by daniel-beck** [comment]
@recena 

> I'm trying to avoid a *big* PR

Me, months ago:

> Make it a new commit, or a separate PR, or a PR to a plugin if you're worried about reviewability, but do *something* with the new layout you're adding here.

**Comment by recena** [comment]
> if you're worried about reviewability

I'm worry about improving the UI of this project 😄

**Comment by oleg-nenashev** [comment]
I have a use-case for it: adding a blank layout (one-column, no header and footer). I used it a couple of times for IFraming of particular Jenkins pages. Maybe it could be a valid case

**Comment by daniel-beck** [comment]
I don't know why this appears to be stalled. To clarify, I just pointed out that

> I'll send a new PR where you will be able to see the usage.

is exactly what I was asking for the entire time.

**Comment by oleg-nenashev** [comment]
Maybe I will provide the usage in the core later. Reassigning to myself

**Comment by oleg-nenashev** [comment]
@recena Should I go forward with the reference impl? During the private chat on Dec you said you will do it on your own, but it has not happened yet

**Comment by recena** [comment]
@oleg-nenashev You are totally right. From now on, I believe I will have more spare time for my personal contribution. Please, give me an one chance more. Ok?

**Comment by recena** [comment]
@daniel-beck As part of this PR I've re-visited how the Setup Wizard implements the full-screen concept. Now, it uses what this PR propones.

File: core/src/main/resources/jenkins/install/SetupWizard/setupWizardFirstUser.jelly [hunk 1]
```diff
@@ -1,55 +1,52 @@
-<!--
-  This is used to create the first user.
--->
 <?jelly escape-by-default='true'?>
-<l:html norefresh="true" title="${it.displayName}" xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:s="/lib/form">
 
-  <l:main-panel>
+  <l:layout norefresh="true" type="full-screen" title="${it.displayName}">
```
Comment (recena) [inline core/src/main/resources/jenkins/install/SetupWizard/setupWizardFirstUser.jelly]: How this step of the Setup Wizard is managed (by using a `iframe`) needs to be changed.

File: core/src/main/resources/jenkins/install/SetupWizard/setupWizardFirstUser.jelly [hunk 2]
```diff
@@ -1,55 +1,52 @@
...
-      <script>
-        $('username').focus();
-      </script>
-    </form>
-  </l:main-panel>
-</l:html>
+        }
+      </style>
+
+     <div id="create-admin-user">
```
Comment (recena) [inline core/src/main/resources/jenkins/install/SetupWizard/setupWizardFirstUser.jelly]: More robust while removing the usage of `iframe`

File: core/src/main/resources/lib/layout/html.jelly [hunk 3]
```diff
@@ -1,171 +0,0 @@
-<!--
```
Comment (recena) [inline core/src/main/resources/lib/layout/html.jelly]: This file only had sense for the Setup Wizard and was a copy of `layout.jelly`. It was created for implementing the full screen layout. Now, different layouts can be managed through `layout.jelly`.

**Comment by recena** [comment]
@oleg-nenashev @daniel-beck Waiting for your final thoughts.

File: core/src/main/resources/jenkins/install/SetupWizard/proxy-configuration.jelly [hunk 4]
```diff
@@ -1,6 +1,8 @@
 <?jelly escape-by-default='true'?>
-<l:html norefresh="true" title="${it.displayName}" xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
-  <l:main-panel>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:s="/lib/form">
```
Comment (kzantow) [inline core/src/main/resources/jenkins/install/SetupWizard/proxy-configuration.jelly]: Need `xmlns:f="/lib/form"`

**Review by kzantow — CHANGES_REQUESTED** [review]

File: core/src/main/resources/jenkins/install/SetupWizard/proxy-configuration.jelly (hunk 4 above)
Comment (recena) [inline core/src/main/resources/jenkins/install/SetupWizard/proxy-configuration.jelly]: Great catch. Is this jelly used currently?

**Comment by recena** [comment]
Unit tests have passed.

**Comment by recena** [comment]
@kzantow Are the changes fine for you?

File: core/src/main/resources/jenkins/install/SetupWizard/setupWizardFirstUser.jelly (hunk 1 above)
Comment (kzantow) [inline core/src/main/resources/jenkins/install/SetupWizard/setupWizardFirstUser.jelly]: why, exactly?

**Review by kzantow — APPROVED** [review]
👍  LGTM

File: core/src/main/resources/jenkins/install/SetupWizard/setupWizardFirstUser.jelly (hunk 1 above)
Comment (recena) [inline core/src/main/resources/jenkins/install/SetupWizard/setupWizardFirstUser.jelly]: @kzantow I think we should not use `iframes`. I would like to work on a proposal based on a wizard (flow) with _real_ steps / states, where we can have a URL per state.

**Comment by daniel-beck** [comment]
Tons of unnecessary whitespace/line break changes even in `?w=1` diff.

**Comment by daniel-beck** [comment]
Regression in the layout of the *Configure Proxy* dialog (no margin), possibly others. I don't like that (based on previous comments) this apparently wasn't tested?

> screen shot

> screen shot

**Comment by recena** [comment]
@daniel-beck Good catch.

Precisely this step needs also to be revisited:

1. Different space (margin) configuration in comparison with other steps
1. Different font configuration

File: core/src/main/resources/lib/layout/layout.jelly [hunk 5]
```diff
@@ -53,7 +53,7 @@ THE SOFTWARE.
       (The permission will be checked against the "it" object.)
     </st:attribute>
     <st:attribute name="type" use="optional">
-      Available values: two-column (by default) or one-column (full-width size).
+      Available values: two-column (by default), one-column (full-width size) or full-screen.
```
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/layout.jelly]: Should mention which versions introduced these options, with the new one being `(since TODO)`

Comment (recena) [inline core/src/main/resources/lib/layout/layout.jelly]: @daniel-beck I can do it but I never saw this annotation of jelly tag.

File: core/src/main/resources/lib/layout/layout.jelly [hunk 6]
```diff
@@ -53,7 +53,7 @@ THE SOFTWARE.
       (The permission will be checked against the "it" object.)
     </st:attribute>
     <st:attribute name="type" use="optional">
-      Available values: two-column (by default) or one-column (full-width size).
+      Available values: two-column (by default), one-column (full-width size) or full-screen (since TODO).
```
Comment (recena) [inline core/src/main/resources/lib/layout/layout.jelly]: `since TODO` was asked by @daniel-beck

**Comment by recena** [comment]
@daniel-beck ping?

**Comment by daniel-beck** [comment]
I think this is the wrong approach (couldn't `/lib/layout` just inherit from from `/lib/html` for headers etc?), but now there's nothing obviously broken anymore AFAICT.

**Comment by daniel-beck** [comment]
To clarify my previous comment, I'm not blocking this. My concerns were addressed. I should have written…

> I'm not convinced this is the right approach.

My review position is therefore ±0.

**Comment by recena** [comment]
@daniel-beck thanks.

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
I disagree with removal of `lib/layout/html.jelly ` in such way. Even if it does not make sense for other usages (proof?), it was a part of the Jenkins API for Jelly, which is documented in Jenkins TagLib documentation. This TagLib site is not being hosted on jenkins.io now (hope it changes at some point), but IMHO it is not safe to just remove it.

I would propose to keepthis Jelly tag somehow. It is fine to mark it as deprecated in the documentation if you do not want to inherit from it like @daniel-beck proposed.

**Comment by recena** [comment]
THIS IS INCREDIBLE!

**Comment by recena** [comment]
@oleg-nenashev `html.jelly` was introduced by @kzantow for providing a full screen layout. This tag is being used only in the setup wizard.

**Comment by recena** [comment]
@oleg-nenashev Curiously, the author of this tag agrees with the changes proposed here.

**Comment by recena** [comment]
I'm remembering this one https://github.com/jenkinsci/jenkins/pull/1717, with more than **122** comments.

**Comment by oleg-nenashev** [comment]
I and @recena discussed this issue internally. I still disagree with such ranting and I still feel strong about my feedback.

I propose the following ways to proceed...

1) Get more votes for @jenkinsci/code-reviewers . I do not veto this PR (nobody can)
2) Decouple the `html.jelly` deletion to a separate PR an continue the discussion there

**Comment by recena** [comment]
If _ranting_ means sharing / expressing my frustration, yes. Each change related with the UI involves a BIG challenge. We should think / reflect why the UI and UX has been one of the most negative part of this project.

**Comment by recena** [comment]
@oleg-nenashev Given the two options, I've choose the second one.

**Comment by recena** [comment]
@oleg-nenashev 

> I would propose to keepthis Jelly tag somehow. It is fine to mark it as deprecated in the documentation if you do not want to inherit from it like @daniel-beck proposed.

This was what I did.

**Review by oleg-nenashev — APPROVED** [review]
👍 for the code itself. Ready to merge assuming it was manually tested

**Comment by oleg-nenashev** [comment]
@daniel-beck wdyt?

**Comment by recena** [comment]
@oleg-nenashev I'm goint to prepare the PR for removing `html.jelly`, as you suggested.

**Comment by recena** [comment]
@oleg-nenashev this was tested many times.

Today, I did smoke tests against: `2.32.3`, `2.41.1` and `master`.

**Comment by recena** [comment]
By the way, I did a rebase with `master, this was my last commit.

**Comment by oleg-nenashev** [comment]
@recena OK, ty. Will sync-up on the release of this fix with Daniel

**Comment by oleg-nenashev** [comment]
According to the feedback from @daniel-beck , he is neutral regarding this change.
Merging according to the feedback from others

**Comment by oleg-nenashev** [comment]
My bad, merged it without squash. But it is not a backporting candidate anyway

**Comment by recena** [comment]
@oleg-nenashev In general, I prefer to merge this kind of changes by using squash mode. But as you mentioned, this is not a candidate for backporting.

**Comment by recena** [comment]
@kzantow @oleg-nenashev @daniel-beck I appreciate a lot your time and reviews.

**Pull request merged**